be/src/storage/rowset/dict_page.cpp

```diff
@@ -0,0 +1,282 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/rowset/dict_page.h"
+
+#include <memory>
+#include <vector>
+
+#include "common/logging.h"
+#include "gutil/casts.h"
+#include "gutil/strings/substitute.h" // for Substitute
+#include "storage/chunk_helper.h"
+#include "storage/range.h"
+#include "storage/rowset/bitshuffle_page.h"
+#include "util/slice.h" // for Slice
+#include "util/unaligned_access.h"
+
+namespace starrocks {
+
+using strings::Substitute;
+
+template <LogicalType Type>
+DictPageBuilder<Type>::DictPageBuilder(const PageBuilderOptions& options)
+        : _options(options),
+          _finished(false),
+          _data_page_builder(nullptr),
+          _dict_builder(nullptr),
+          _encoding_type(DICT_ENCODING) {
+    // initially use DICT_ENCODING
+    _data_page_builder = std::make_unique<BitshufflePageBuilder<DataTypeTraits<Type>::type>>(options);
+    _data_page_builder->reserve_head(BINARY_DICT_PAGE_HEADER_SIZE);
+    PageBuilderOptions dict_builder_options;
+    dict_builder_options.data_page_size = _options.dict_page_size;
+    _dict_builder = std::make_unique<BitshufflePageBuilder<Type>>(dict_builder_options);
+    reset();
+}
+
+template <LogicalType Type>
+bool DictPageBuilder<Type>::is_page_full() {
+    if (_data_page_builder->is_page_full()) {
+        return true;
+    }
+    return _encoding_type == DICT_ENCODING && _dict_builder->is_page_full();
+}
+
+template <LogicalType Type>
+uint32_t DictPageBuilder<Type>::add(const uint8_t* vals, uint32_t count) {
+    if (_encoding_type == DICT_ENCODING) {
+        DCHECK(!_finished);
+        DCHECK_GT(count, 0);
+        ValueCodeType value_code = -1;
+        // Manually devirtualization.
+        auto* code_page = down_cast<BitshufflePageBuilder<DataTypeTraits<Type>::type>*>(_data_page_builder.get());
+
+        if (_data_page_builder->count() == 0) {
+            memcpy(&_first_value, vals, sizeof(ValueType));
+        }
+
+        for (int i = 0; i < count; ++i) {
+            ValueType value = *reinterpret_cast<const ValueType*>(vals + i * SIZE_OF_TYPE);
+            auto iter = _dictionary.find(value);
+            if (iter != _dictionary.end()) {
+                value_code = iter->second;
+            } else if (_dict_builder->add(vals + i * SIZE_OF_TYPE, 1) > 0) {
+                value_code = _dictionary.size();
```

Comment:
use value instead of vals + i * SIZE_OF_TYPE
it be better if  dict_builder->add method's paramter type adopts const T& instead of T*